### ISSUE TITLE
Formatting adds erroneous comma at end of struct update syntax when missing comma before ..

### ISSUE THREAD

**Issue by alice-i-cecile** [rust-lang/rustfmt#5604]
Following up on https://github.com/rust-lang/rust/issues/104373, when a comma is missed before the `..Default::default()`, rust format will then "helpfully" add a comma after the `..Default::default()`.

When you fix the error with the missing comma, you now have a different error ("no trailing commas after...") that the user did not make, as automatic formatting added a comma in its confusion. 

I'm new to this project, but the new error enum variant proposed in https://github.com/rust-lang/rust/pull/104504 may be helpful to identify these cases.

**Comment by ytmimi** [comment]
Thanks for reaching out. Using the latest master `rustfmt 1.5.1-nightly (ee2bed96 2022-11-08)` this is what I'm getting.


**Input**
```rust
fn main(){
    Outer {
        inner: Inner {
            a: 1,
            b: 2,
        }
        ..Default::default()
    };
}
```

**Output**
```rust
fn main() {
    Outer {
        inner: Inner { a: 1, b: 2 }..Default::default(),
    };
}
```

From my understanding the expected output is something like:

```rust
fn main() {
    Outer {
        inner: Inner { a: 1, b: 2 },
        ..Default::default(),
    };
}
```

**Comment by alice-i-cecile** [comment]
If we want to stick to pure formatting, the input shouldn't change at all. But if you want to automatically fix this class of mistake (which would be super slick, but maybe too implicit?), the expected output is actually

```rust
fn main() {
    Outer {
        inner: Inner { a: 1, b: 2 },
        ..Default::default()
    };
}
```

**Comment by TehPers** [comment]
Is it possible that the issue is that the input looks like range syntax?

**Comment by ytmimi** [comment]
@TehPers your hunch is correct. I just dug into this and from the perspective of the AST, which rustfmt operates on, `Inner { a: 1, b: 2 }..Default::default()`, gets parsed as an [`ast::Expr`](https://doc.rust-lang.org/beta/nightly-rustc/rustc_ast/ast/struct.Expr.html) with a kind of [ExprKind::Range](https://doc.rust-lang.org/beta/nightly-rustc/rustc_ast/ast/enum.ExprKind.html#variant.Range).

Sadly, I don't think there's an easy fix for this on the rustfmt side 😕

**Comment by TehPers** [comment]
I agree. I think there's some ambiguity that's preventing this. For example, how is the following supposed to be formatted?

```rs
#[derive(Default)]
struct Inner;

struct Outer {
    inner: Range<Inner>,
}

fn main() {
    let outer = Outer {
        inner: Default::default() // is this supposed to be Range<Inner> or Inner? compiles as Inner
        ..Default::default()
    };
}
```

**Comment by ytmimi** [comment]
Although I'm not sure that rustfmt is the right place to implement a fix for this, if anyone's interested in coming up with a solution this would be a good place to start:

https://github.com/rust-lang/rustfmt/blob/ee2bed96d60fd7e46b1fb868f6a8f27e3a8058d0/src/expr.rs#L269-L342

**Comment by ytmimi** [comment]
@alice-i-cecile given that https://github.com/rust-lang/rust/issues/104373 is closed can this be closed too? As mentioned above, the reason why rustfmt produces the formatting that it does is a result of how the rustc_parser parses `expr..Default::defaut()` as a range when the comma is missing. I'm not as familiar with the parser, but I'm assuming a proper fix for this would be to ensure that the rustc_parser parses `..Default::default()` as [`ast::StructRest::Base`](https://doc.rust-lang.org/beta/nightly-rustc/rustc_ast/ast/enum.StructRest.html) even when the trailing comma is missing.

**Comment by ytmimi** [comment]
If we do want to try to special case the formatting on the rustfmt side it might be better to look into [`rewrite_struct_lit`](https://github.com/rust-lang/rustfmt/blob/ee2bed96d60fd7e46b1fb868f6a8f27e3a8058d0/src/expr.rs#L1531) instead of looking into the code that's responsible for rewriting ranges.
